File: ruby/validation.rb
def file_is_valid(filetext)
  is_valid = has_at_least_one_piece(filetext) &&
    valid_number_of_lines(filetext) &&
    valid_chars(filetext) &&
    valid_width(filetext) &&
    valid_number_of_chars(filetext) &&
    valid_shapes(filetext)

  return is_valid
end

def valid_chars(filetext)
  i = 1
  filetext.each_line do |line|
    if i % 5 == 0
      return false if line != "\n"
    else
      line.chars.each { |char| return false unless ["\n", "#", "."].include? char }
    end
    i += 1
  end
end

def valid_width(filetext)
  i = 1
  filetext.each_line do |line|
    if i % 5 == 0
      return false if line.chars.count != 1
    else
      return false if line.chars.count != 5
    end
    i += 1
  end
  return true
end

def valid_number_of_chars(filetext)
  i = 1
  hashtags = 0
  points = 0

  filetext.each_line do |line|
    if i % 5 == 0
      hashtags = 0
      points = 0
    end

    line.chars.each do |char|
      hashtags += 1 if char == "#"
      points += 1 if char == "."
    end

    if i % 5 == 4
      return false if hashtags != 4 || points != 12
    end
    i += 1
  end
  return true
end

def valid_shapes(filetext)
  filetext.blocks.each do |block_str|
    # read block into an array of arrays
    hashtags_with_two_neighbors = 0
    block = block_string_to_array(block_str)
    for i in 0..3
      for j in 0..3
        if block[i][j] == "#"
          hashtags_with_two_neighbors += 1 if number_of_neighbors(block, i, j) == 2
        end
      end
    end
    return false if hashtags_with_two_neighbors < 2
  end
  return true
end

def has_at_least_one_piece(filetext)
  number_of_lines(filetext) >= 5
end

def number_of_neighbors(block, i, j)
  number_of_neighbors = 0
  directions = [[0,1], [1,0], [-1,0], [0,-1]]
  directions.each do |direction|
    if (0..3).to_a.include?(i + direction[0]) && (0..3).to_a.include?(j + direction[1])
      if block[i + direction[0]][j + direction[1]] == "#"
        number_of_neighbors += 1
      end
    end
  end
  return number_of_neighbors
end

def valid_number_of_lines(filetext)
  number_of_lines(filetext) % 5 == 0
end

def number_of_lines(filetext)
  filetext.lines.count + 1
end

File: validation.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   validation.c                                       :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: fwuensch <fwuensch@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/01/10 22:02:28 by ypetitje          #+#    #+#             */
/*   Updated: 2019/01/20 19:24:07 by fwuensch         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "fillit.h"

int		file_is_valid(char *filetext, char **lines)
{
	int		is_valid;

	// printf("has_at_least_one_piece: %d\n", has_at_least_one_piece(lines));
	// printf("valid_number_of_lines: %d\n", valid_number_of_lines(lines));
	// printf("valid_chars: %d\n", valid_chars(lines));
	// printf("valid_width: %d\n", valid_width(lines));
	// printf("valid_number_of_chars: %d\n", valid_number_of_chars(lines));

	is_valid = (has_at_least_one_piece(lines) &&
		valid_number_of_lines(lines) &&
		valid_chars(lines) &&
		valid_width(lines) &&
		valid_number_of_chars(lines));
	filetext[0] = filetext[0];
	return (is_valid);
}

int		has_at_least_one_piece(char **lines)
{
	int		i;

	i = 0;
	while (lines[i] != NULL)
	{
		i++;
	}
	return (i + 1 >= 5);
}

int		valid_number_of_lines(char **lines)
{
	int		i;

	i = 0;
	while (lines[i] != NULL)
	{
		i++;
	}
	return ((i + 1) % 5 == 0);
}

int		valid_chars(char **lines)
{
	int		i;
	int		j;
	int		curr_line;

	i = 1;
	curr_line = 0;
	while(lines[curr_line])
	{
		if (i % 5 == 0)
		{
			if (ft_strlen(lines[curr_line]) != 0)
				return (0);
		}
		else
		{
			j = 0;
			while (lines[curr_line][j])
			{
				if (lines[curr_line][j] != '.' && lines[curr_line][j] != '#')
					return (0);
				j++;
			}
		}
		i++;
		curr_line++;
	}
	return (1);
}

int		valid_width(char **lines)
{
	int		i;
	int		curr_line;

	i = 1;
	curr_line = 0;
	while(lines[curr_line])
	{
		if (i % 5 == 0)
		{
			if (ft_strlen(lines[curr_line]) != 0)
				return (0);
		}
		else
		{
			if (ft_strlen(lines[curr_line]) != 4)
				return (0);
		}
		i++;
		curr_line++;
	}
	return (1);
}

int		valid_number_of_chars(char **lines)
{
	int		i;
	int		j;
	int		hashtags;
	int		points;
	int		curr_line;

	i = 1;
	hashtags = 0;
	points = 0;
	curr_line = 0;
	while(lines[curr_line])
	{
		if (i % 5 == 0)
		{
			hashtags = 0;
			points = 0;
		}
		j = 0;
		while (lines[curr_line][j])
		{
			if (lines[curr_line][j] == '#')
				hashtags++;
			if (lines[curr_line][j] == '.')
				points++;
			j++;
		}
		if (i % 5 == 4)
		{
			if ((hashtags != 4) || (points != 12))
				return (0);
		}
		i++;
		curr_line++;
	}
	return (1);
}

// int		valid_shapes(char **lines)
// {
// 	int		block;
// 	int		i;
// 	int		j;
// 	int		hashtag_with_two_neighbors;

// 	block = 0;
// 	while (lines[block * 5])
// 	{
// 		i = 0;
// 		while (i < 4)
// 		{
// 			while (j < 4)
// 			{
// 				if (lines[i][j] == '#')
// 					if (number_of_neighbors(block))
// 						hashtag_with_two_neighbors++;
// 				j++;
// 			}
// 			i++;
// 		}


// 	}

// }
